### PR TITLE
udevadm/hwdb: Return non-zero exit code on error

### PR DESCRIPTION
**Pull request by drinkcat**
It seems useful to return a non-zero exit code when there is a parsing error in one of the hwdb file.

This is useful on Chromium OS as we run `udevadm hwdb --update` during our image creation stage (see http://crbug.com/823605 for details), and we want the build to fail if the hwdb is invalid.

### CODE REVIEW

**Comment by drinkcat** [comment]
IIUC, it looks like `systemd-nspawn smoke test` failed, which should not be affected by this change?!

```
TEST RUN: systemd-nspawn smoke test [FAILED]
Makefile:4: recipe for target 'run' failed
make: Leaving directory '/tmp/autopkgtest.RH13ar/build.VRZ/debian/test/TEST-13-NSPAWN-SMOKE'
make: *** [run] Error 1
cp: -r not specified; omitting directory '/var/tmp/systemd-test.dpySLs/journal/324196b8a7d74e758e6d21a7aaf8a88c'
autopkgtest [02:07:50]: test upstream: -----------------------]
autopkgtest [02:07:50]: test upstream:  - - - - - - - - - - results - - - - - - - - - -
upstream             FAIL non-zero exit status 1
```

File: src/udev/udevadm-hwdb.c [hunk 1]
```diff
@@ -646,9 +651,12 @@ static int adm_hwdb(struct udev *udev, int argc, char *argv[]) {
                 }
                 STRV_FOREACH(f, files) {
                         log_debug("reading file '%s'", *f);
-                        import_file(udev, trie, *f);
+                        if (import_file(udev, trie, *f) < 0)
+                            rc = EXIT_FAILURE;
                 }
                 strv_free(files);
+                if (!rc)
```
Comment (keszybz) [inline src/udev/udevadm-hwdb.c]: This condition seems to be reversed.

Also, please use an explicit comparison, e.g. `rc != 0`.

File: src/udev/udevadm-hwdb.c [hunk 2]
```diff
@@ -461,6 +461,7 @@ static int import_file(struct udev *udev, struct trie *trie, const char *filenam
         FILE *f;
         char line[LINE_MAX];
         struct udev_list match_list;
+        int ret = 0;
```
Comment (keszybz) [inline src/udev/udevadm-hwdb.c]: `r` would be better than `ret`, because we use `r` for this role elsewhere.

File: src/udev/udevadm-hwdb.c (hunk 1 above)
Comment (drinkcat) [reply src/udev/udevadm-hwdb.c]: It's correct. But replaced by `rc != EXIT_SUCCESS`.

File: src/udev/udevadm-hwdb.c (hunk 2 above)
Comment (drinkcat) [reply src/udev/udevadm-hwdb.c]: Done.

**Comment by drinkcat** [comment]
> I don't think we can do this change. Right now, if there's some typo in one of the files, we log a warning but process the rest of the file and other files. "Soft degradation". With the proposed change, a typo in any file would disable hwdb updates completely.

I updated the patch to do full soft degradation (it was already continuing to read the next file, but stopped at the first error within a file).

With this, the only difference is that the code returns a non-zero error code when anything goes wrong during parsing. Do you think that's ok? Or should we still add a `--strict` flag?

> You could make this optional, e.g. with a new --strict switch. But maybe a different approach would be better: we have hwdb/parse_hwdb.py which contains a much stricter parser for this, with additional checks. We use it for testing, and it is also used in libinput. You could try running that as part of your pipeline.

It seems like `parse_hwdb.py` is too strict ,-( For example, it fails on `./hwdb/20-acpi-vendor.hwdb`:
```
WARNING: evdev is not available
Cannot parse 20-acpi-vendor.hwdb: Expected stringEnd (at char 188), (line:9, col:1)
20-acpi-vendor.hwdb: 0 match groups, 0 matches, 0 properties
```

**Comment by keszybz** [comment]
parse_hwdb.py has rules only for `hwdb/[67]0-*.hwdb`. It could be extended to those other files. It wasn't mostly because they are really simple and generated automatically and hwdb-parse is fairly slow.

File: src/udev/udevadm-hwdb.c [hunk 3]
```diff
@@ -531,19 +536,23 @@ static int import_file(struct udev *udev, struct trie *trie, const char *filenam
...
 
-                        insert_data(trie, &match_list, line, filename);
+                        err = insert_data(trie, &match_list, line, filename);
+                        if (err < 0)
+                                r = err;
                         break;
                 };
         }
 
+  out:
```
Comment (keszybz) [inline src/udev/udevadm-hwdb.c]: Hmm, why is this needed?

File: src/udev/udevadm-hwdb.c [hunk 4]
```diff
@@ -646,9 +655,12 @@ static int adm_hwdb(struct udev *udev, int argc, char *argv[]) {
                 }
                 STRV_FOREACH(f, files) {
                         log_debug("reading file '%s'", *f);
-                        import_file(udev, trie, *f);
+                        if (import_file(udev, trie, *f) < 0)
+                            rc = EXIT_FAILURE;
```
Comment (keszybz) [inline src/udev/udevadm-hwdb.c]: If `import_file` fails, you jump to `out`, so hwdb.bin is **not** written.

Please use 8 spaces of indentation everywhere (see CODING_STYLE).

Comment (drinkcat) [reply src/udev/udevadm-hwdb.c]: Right, of course, sorry about that. Fixed.

File: src/udev/udevadm-hwdb.c (hunk 3 above)
Comment (drinkcat) [reply src/udev/udevadm-hwdb.c]: Not needed, dropped.

**Comment by keszybz** [comment]
OK, so now the code looks good and does what the patch description says.

> With this, the only difference is that the code returns a non-zero error code when anything goes wrong during parsing. Do you think that's ok? Or should we still add a --strict flag?

Yes, a flag for the new behaviour is still needed.

(A longer explanation: we had similar issues with tmpfiles, sysusers, etc, where it turns out that if stuff is being run automatically at boot or during upgrades, it cannot fail on local mistakes in configuration. At the size of the whole distro, there's always one package or another that installs some file with a bug, and services start failing and people file bugs, etc. What works is to report success, but provide clear warnings in the journal. Then people can see the error and fix it, without users thinking that stuff is broken.)

**Comment by drinkcat** [comment]
> Yes, a flag for the new behaviour is still needed.

Makes sense, thanks.

Patch updated.

File: src/udev/udevadm-hwdb.c [hunk 5]
```diff
@@ -568,6 +576,7 @@ static int adm_hwdb(struct udev *udev, int argc, char *argv[]) {
         static const struct option options[] = {
                 { "update",  no_argument,       NULL, 'u'     },
                 { "usr",     no_argument,       NULL, ARG_USR },
+                { "strict",  no_argument,       NULL, 's'     },
```
Comment (keszybz) [inline src/udev/udevadm-hwdb.c:569]: This also needs to be added to help() above.

Comment (drinkcat) [reply src/udev/udevadm-hwdb.c:569]: Done.

**Comment by drinkcat** [comment]
> Unforatunately, I completely forgot that udevadm hwdb is the deprecated interface, and systemd-hwdb update is the new thing. So the same logic update should be done in src/hwdb/hwdb.c, and the new option for systemd-hwdb should be documented in man/systemd-hwdb.xml.

Right, it just happens that we still use the old interface ,-)

Anyway, added a second commit to fix systemd-hwdb as well.

Thanks!

File: src/hwdb/hwdb.c [hunk 6]
```diff
@@ -684,18 +693,23 @@ static int hwdb_update(int argc, char *argv[], void *userdata) {
...
-        return label_fix(hwdb_bin, false, false);
+        return r;
 }
 
 static void help(void) {
         printf("Usage: %s OPTIONS COMMAND\n\n"
                "Update or query the hardware database.\n\n"
                "  -h --help            Show this help\n"
                "     --version         Show package version\n"
+               "  -s --strict          When updating, return non-zero exit value on any parsing error\n"
```
Comment (keszybz) [inline src/hwdb/hwdb.c]: Oh, somehow this fell off my radar. Sorry for the delay.

--help output needs to be limited to 80 columns. We actually have a test for this, which should have caught this (this line is 86 characters wide now).

Comment (keszybz) [reply src/hwdb/hwdb.c]: Oh, all the CI is failing because of the width issue:
```
252/257 check-help-systemd-hwdb                 FAIL     0.02 s
```

**Comment by keszybz** [comment]
A rebase is also needed.

File: src/hwdb/hwdb.c (hunk 6 above)
Comment (drinkcat) [reply src/hwdb/hwdb.c]: Fixed. Thanks.

**Comment by keszybz** [comment]
Thanks!

**Comment by drinkcat** [comment]
Thanks to you!
